### DISCUSSION TITLE
Slideshow support for custom logos

### DISCUSSION THREAD

**Discussion by Stat-Mat** [frntc/Sidekick64#50] [Ideas]
Would it perhaps be possible to support a slideshow feature for the custom logos when used with CRT and PRG files? e.g. the format to use for the filename could perhaps be [GAME_NAME]-[X].tga, where [X] is the numerical index (e.g. Airwolf-1.tga, Airwolf-2.tga etc).

I'm not sure if there's enough spare processing power to support this whilst a game is running (I'm guessing probably not given the criticality of timings), but it'd certainly be nice to support it if possible. For example, in OneLoad64 I was thinking I could provide both the loading screen if present followed by an in-game screenshot and then the custom logo would cycle between the two. Perhaps the duration could be defaulted to 5 seconds, but also be configurable in the config with a minimum value of 5 (i.e. if set lower, then ignored)?

**Comment by frntc** [comment]
I'll think about it -- having more "action" on the TFT is always a good idea. So far it was rather the content lacking :-)

For PRGs there is no timing critical thing: after it has been transferred, the RPi is basically idling (unless you use the SID-emulation).

For CRTs it depends, but I think it should be possible. It's probably just diligence work. What CRT-formats do you have (in mind/in the collection)? (MagicDesk, EF, ...?)

**Reply by Stat-Mat** [reply]
Thanks for the consideration. Indeed, it would be nice to see a little more made of the TFT. ;)

The core CRT single-load games in OneLoad64 are in Magic Desk format, whereby the game is quickly copied into the C64s RAM before the cart is banked out, so I guess these should probably be fine - i.e. display the first image as you do now before starting the cart emulation, and by the time the next image is needed, the cart isn't being accessed. However, I appreciate this is rather specific to these conversions and may not be suitable for all carts.

**Reply by frntc** [reply]
if it's MD and copy-to-RAM it'll be easy once banked out. The most complicated thing might be searching for/loading the images :)

Would it be an option to create TGA with width=240 and height=240*N where N is the number of images? (I'd have to check my code, but this might be the most transparent solution of handling single images vs. slide shows in the end)

Btw. don't expect stunning frame rates, as the transfer has some overhead and is in lockstep with the clock speed. It was quite some work getting the animation for the SID-emulation at decent speed...

**Reply by Stat-Mat** [reply]
Interesting idea, indeed it would be easier to load one file which has the TGA files stacked as you suggest. It would be easy for me to create the combined images, as I am generating the TGA files with my own C# code.

And sure, don't worry, not expecting fast blitting to the TFT or anything. ;)

**Reply by frntc** [reply]
I was thinking a bit about the slideshow. To ensure backwards compatibility, how about this: there is the .tga (as before) which serves as the "loading screen". If there is another .tga (according to a specific naming scheme) which has a height 240*N then it will be shown as a slideshow once the PRG has started, or the CRT is deactivated?

**Reply by Stat-Mat** [reply]
That's a nice idea to maintain backwards compatibility and I think that could work well. We just need to decide on the naming scheme for the second file. Perhaps just the suffix "-slideshow" (e.g. Airwolf-slideshow.tga)? Oh, and can I assume the first file will still be included in the slidehow when looping?

**Reply by frntc** [reply]
I prepared that PRG and MagicDesk launchers are looking for a "example-slideshow.tga" (240 * 240 * N size), and also for (even more optional) "example-slideshow.time", which stores 1 byte per image in the slideshow to set the display time (in 1/10th seconds). Would you need anything else in this regard?

**Reply by Stat-Mat** [reply]
Many thanks for your work on this. I guess I just need to clarify the following:

1. Will the initial loading screen image be included in the slideshow when looping around, or would it be best to simply tack another copy of it onto the end of the "example-slideshow.tga" (240 * 240 * N size) file?
2. I'm not sure 100ms display time increments are necesary, or differing display times between slides (I wasn't intending to use it for animation for example) - sorry if I was unclear on that. Wouldn't it just be simpler to have a global value for this in the config which perhaps defaults to 5 seconds (specified in ms for increased flexibility)?

**Comment by frntc** [comment]
1) no it's kept separate. If you want it to be included, it has to be part of the 240*240*N image.
2) sure, 100ms is a bit fine-grained, in particular given that fading from one image to another already takes ~1s or so :) (i.e. don't think of animations!) If the ".time"-file is not provided, it will take a global (fixed) value.

**Reply by Stat-Mat** [reply]
Cool, that certainly makes sense. Sounds good to me. :)

**Reply by Stat-Mat** [reply]
Ah, one thing I've just realised is that I was thinking that each slideshow file would simply be a concatenation of the required TGA files (i.e. each would still be 240x240 in resolution with the TGA header at the start of each file. However, were you thinking that all the bitmaps would be merged into a single file? After re-reading our discussion above, I'm pretty sure this is what you were thinking. i.e. if the slideshow file contained two files, then it would be a single 240x480 TGA file.

**Reply by frntc** [reply]
yes, exactly (e.g. 240x480). I think it's better to go this way, as searching for and reading many files from SD-card is slow.

**Reply by Stat-Mat** [reply]
Okay, understood. I will adapt my logic accordingly. ;)

**Reply by Stat-Mat** [reply]
Okay, you can grab an example of the files created with my updated script for Operation Wolf for your testing (contains PRG, CRT, as well as single and slideshow images) [[here]](https://drive.google.com/file/d/1Uf04RX4TZE2XkXYC6PkmS-2-kpfOqo7Y/view?usp=share_link)

**Comment by StmLord** [comment]
Is that future implemented yet? 
I'm creating a collection of CRT logos for SK64 and for collections I could create slideshows with every game in it (although some may not be possible as having 20+ games in a collection :D )

**Reply by Stat-Mat** [reply]
Yes, frenetic very kindly implemented this feature and it is included in the latest firmware when used in conjunction with Magic Desk cartridges (https://github.com/frntc/Sidekick64/releases/tag/v0.51c). Additionally, the OneLoad64 Games Collection v4 now includes slideshow images for the SK64 including loading screens and in-game screenshots. I actually wrote a tool to help create these from larger images, so I could possibly tweak and share that if it's useful. :)

**Reply by StmLord** [reply]
hmm, is it only for magic desk carts? Not for Easyflash that are the most collections?
I have already created logos for about 60 releases I think there are not many more, I left out almost all collections so I can create slideshow logos for them. I'm wondering where to upload the logos collection when I'm done, any thoughts?

**Reply by frntc** [reply]
to be honest, I have to look into the code :-)

In any case, the slide shows only begin when the cartridge is deactivated (less hassle with cache preloading, which is difficult). Most EF CRTs do not deactivate, but access different banks all the time.
